scalyr_agent/copying_manager/worker.py

```diff
@@ -1036,21 +1043,58 @@ def __scan_for_new_bytes(self, current_time=None):
         for processor in self.__log_processors:
             processor.scan_for_new_bytes(current_time)
 
-    def get_full_checkpoints(self):  # type: () -> Dict[six.text_type, Any]
+    def _update_checkpoints(self, current_time):
+        """
+        Update the checkpoints collection from the current state of the log processors.
+        :param current_time:
+        :return:
+        """
+        # get all log processors, including those that need to be removed.
+        log_processors = self.__log_processors
+
+        with self._checkpoints_lock:
+            # remove stale checkpoints using the same config option as for the whole file staleness check.
+            for path, checkpoint in list(self._checkpoints.items()):
+                checkpoint_time = checkpoint.get("time")
+                if checkpoint_time:
+                    if (
+                        current_time
+                        > checkpoint_time + self.__config.max_allowed_checkpoint_age
+                    ):
+                        # remove stale checkpoint states.
+                        del self._checkpoints[path]
+
+            for processor in log_processors:
+                processor_checkpoint = processor.get_checkpoint()
```

Comment:
It seems like we should do a deep copy here and not work with the object returned by the `get_checkpoint` method.